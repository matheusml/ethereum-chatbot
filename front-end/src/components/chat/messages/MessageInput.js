import React from "react";
import {
  StyleSheet,
  Text,
  View,
  TextInput,
  KeyboardAvoidingView,
  TouchableOpacity
} from "react-native";
import PropTypes from "prop-types";

import dismissKeyboard from "dismissKeyboard";

export default class MessageInput extends React.Component {
  constructor(props) {
    super(props);

    this.state = { typing: "" };

    this.sendMessage = this.sendMessage.bind(this);
  }

  sendMessage() {
    const { typing } = this.state;

    const newMessage = {
      message: this.state.typing,
      sender: "You"
    };

    this.setState(
      {
        typing: ""
      },
      () => this.props.sendMessage(newMessage)
    );

    dismissKeyboard();
  }

  render() {
    return (
      <KeyboardAvoidingView behavior="padding">
        <View style={styles.footer}>
          <TextInput
            value={this.state.typing}
            style={styles.input}
            underlineColorAndroid="transparent"
            placeholder="Send a message"
            onChangeText={typing => this.setState({ typing })}
            editable={!this.props.loading}
            selectTextOnFocus={!this.props.loading}
          />
          <TouchableOpacity
            disabled={this.props.loading}
            onPress={this.sendMessage}
          >
            <Text style={styles.send}>Send</Text>
          </TouchableOpacity>
        </View>
      </KeyboardAvoidingView>
    );
  }
}

MessageInput.propTypes = {
  sendMessage: PropTypes.func.isRequired,
  loading: PropTypes.bool
};

const styles = StyleSheet.create({
  footer: {
    flexDirection: "row",
    backgroundColor: "#eee"
  },
  input: {
    paddingHorizontal: 20,
    fontSize: 18,
    flex: 1
  },
  send: {
    alignSelf: "center",
    color: "#0aa6dd",
    fontSize: 16,
    fontWeight: "bold",
    padding: 20
  }
});
