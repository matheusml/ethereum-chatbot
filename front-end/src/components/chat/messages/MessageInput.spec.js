import React from "react";
import dismissKeyboard from "dismissKeyboard";

import MessageInput from "./MessageInput";

import renderer from "react-test-renderer";

jest.mock("dismissKeyboard");

const sendMessageMock = jest.fn();

const props = {
  sendMessage: sendMessageMock,
  loading: false
};

describe("MessageInput", () => {
  it("should match snapshot", () => {
    const tree = renderer.create(<MessageInput {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should match snapshot with loading", () => {
    const tree = renderer
      .create(<MessageInput {...props} loading={true} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should have an initial state", () => {
    const instance = renderer.create(<MessageInput {...props} />).getInstance();
    expect(instance.state).toEqual({ typing: "" });
  });

  describe("sendMessage", () => {
    it("should update the state", () => {
      const instance = renderer
        .create(<MessageInput {...props} />)
        .getInstance();
      const newMessage = {
        message: "new message",
        sender: "You"
      };
      instance.sendMessage(newMessage);
      expect(instance.state).toEqual({ typing: "" });
    });

    it("should invoke dismissKeyboard", () => {
      const instance = renderer
        .create(<MessageInput {...props} />)
        .getInstance();
      const newMessage = {
        message: "new message",
        sender: "You"
      };
      instance.sendMessage(newMessage);
      expect(dismissKeyboard).toHaveBeenCalledWith();
    });

    it("should invoke sendMessage", () => {
      const instance = renderer
        .create(<MessageInput {...props} />)
        .getInstance();
      const newMessage = {
        message: "",
        sender: "You"
      };
      instance.sendMessage(newMessage);
      expect(sendMessageMock).toHaveBeenCalledWith(newMessage);
    });
  });
});
