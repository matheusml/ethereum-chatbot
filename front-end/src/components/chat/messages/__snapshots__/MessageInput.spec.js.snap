// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInput should match snapshot 1`] = `
<View
  keyboardVerticalOffset={0}
  onLayout={[Function]}
  style={
    Array [
      undefined,
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#eee",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={true}
      onChangeText={[Function]}
      placeholder="Send a message"
      selectTextOnFocus={true}
      style={
        Object {
          "flex": 1,
          "fontSize": 18,
          "paddingHorizontal": 20,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "lightseagreen",
            "fontSize": 16,
            "fontWeight": "bold",
            "padding": 20,
          }
        }
      >
        Send
      </Text>
    </View>
  </View>
</View>
`;

exports[`MessageInput should match snapshot with loading 1`] = `
<View
  keyboardVerticalOffset={0}
  onLayout={[Function]}
  style={
    Array [
      undefined,
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#eee",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={false}
      onChangeText={[Function]}
      placeholder="Send a message"
      selectTextOnFocus={false}
      style={
        Object {
          "flex": 1,
          "fontSize": 18,
          "paddingHorizontal": 20,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "lightseagreen",
            "fontSize": 16,
            "fontWeight": "bold",
            "padding": 20,
          }
        }
      >
        Send
      </Text>
    </View>
  </View>
</View>
`;
